### PR TITLE
jextract: 2023-04-14 -> 2023-11-27 (JDK 21)

### PR DESCRIPTION
**Pull request by msgilligan**
## Description of changes

Upgrade to build with JDK 21 and use the ~~2023-09-25~~ 2023-11-27 commit from the `jdk21` branch of `jextract`. See Issue #270791.

## Things done

~~Nothing except the changes~~. I am new to Nix ~~and don't even know how to test this locally~~. Submitting ~~as a DRAFT PR~~ to attach to Issue #270791.

**Update:** I've installed and tested with a small `jextract`/`j.l.foreign` project on `darwin-aarch64` and it seems to be working.

### CODE REVIEW

**Review by msgilligan — COMMENTED** [review]
OK, I've tested this on aarch64-darwin and it seems to be working.

**Comment by msgilligan** [comment]
> Yes, you should squash it first. I did a basic testing on my x86_64 machine and it seems OK. No other change is required if your testing on darwin goes successfully.

OK, I squashed and force-pushed. I also did some testing on `aarch64-darwin` and it seems to be working.

**Review by SharzyL — APPROVED** [review]

**Comment by msgilligan** [comment]
@SharzyL is there anything else we need to do before this can be merged?

**Comment by msgilligan** [comment]
> I think it is OK now, but I have no merge permission

So there is nothing else we need to do? Just wait for someone with merge permission to do the merge? Is there a way to notify them or assign the issue to them somehow?

File: pkgs/development/tools/java/jextract/default.nix
Comment (jlesquembre) [inline pkgs/development/tools/java/jextract/default.nix:47]: Could we use the latest JDK21 commit? I tested it and looks good: 

```suggestion
    rev = "f8d229775e8260d4bebf21525b3151b9fff128ae";
    hash = "sha256-jkUCh4oSgilszvvU7RpozFATIghaA9rJRAdUIl5jTHM=";
```

Comment (msgilligan) [reply pkgs/development/tools/java/jextract/default.nix:47]: I chose `a3c3644580a5144840de341e7c1cbf7dc4de3206` (`# Tweak build for JDK 21`) because it corresponds with the `jextract` binary released [on jdk.java.net](https://jdk.java.net/jextract/).

The hash you are suggesting (https://github.com/openjdk/jextract/commit/f8d229775e8260d4bebf21525b3151b9fff128ae) has a fix for a Windows test and a fix for one of the code samples so does not affect the `jextract` binary in any way.

There is also this commit https://github.com/openjdk/jextract/commit/8730fcf05c229d035b0db52ee6bd82622e9d03e9 (on `master`) that seems to have some actual fixes.

I'd be happy with any of those three commits. @SharzyL what do you suggest?

Comment (jlesquembre) [reply pkgs/development/tools/java/jextract/default.nix:47]: Since jextract doesn't have any releases yet, I'd go with https://github.com/openjdk/jextract/commit/8730fcf05c229d035b0db52ee6bd82622e9d03e9 since it included a fix. But that's my personal opinion.

Comment (SharzyL) [reply pkgs/development/tools/java/jextract/default.nix:47]: If you find these new changes useful, it is totally fine to include these new changes. Just remember to update the commit message, PR title and do tests. jextract have no release tags hence we can and only can pick commits from master.

Comment (msgilligan) [reply pkgs/development/tools/java/jextract/default.nix]: OK, I've updated to the current `master` of https://github.com/openjdk/jextract and tested on `aarch64-darwin` and it is working!

**Review by SpriteOvO — APPROVED** [review]
LGTM, thanks!

**Review by SharzyL — APPROVED** [review]

**Comment by jlesquembre** [comment]
Result of `nixpkgs-review pr 271127` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>jextract</li>
  </ul>
</details>

**Review by jlesquembre — APPROVED** [review]

**Pull request merged**
